### PR TITLE
NikitK/ Direct Client to Unsubscribe Email without Login

### PR DESCRIPTION
**Pull request by NikitK-deriv**
Changes:

-    unsubscribe page 

## Type of change

-   [ ] Bug fix
-   [x] New feature
-   [ ] Update feature
-   [ ] Refactor code
-   [ ] Translation to code
-   [ ] Translation to crowdin
-   [ ] Script configuration
-   [ ] Improve performance
-   [ ] Style only
-   [ ] Dependency update
-   [ ] Documentation update
-   [ ] Release

### CODE REVIEW

File: src/common/utility.ts [hunk 1]
```diff
@@ -355,7 +355,7 @@ export const removeSpecialCharacterUrl = (url) =>
 export const queryParams = {
     get: (key) => {
         const params = new URLSearchParams(isBrowser() && location.search)
-        const param_values = {}
+        let param_values
```
[the remark this answers is not in the document]
Comment (NikitK-deriv) [reply src/common/utility.ts:358]: Isn't a reassign? `param_values[key] = params.get(k)`

File: src/pages/unsubscribe/index.tsx [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import { Button } from 'components/form'
+import { localize, WithIntl } from 'components/localization'
+import Layout from 'components/layout/layout'
+import CheckIcon from 'images/common/check_icon.png'
+import device from 'themes/device'
+import { useDerivApi } from 'components/hooks/use-deriv-api'
+import { queryParams } from 'common/utility'
+import { decode } from 'common/url-base64-functions'
+
+const UnsubscrubeWrapper = styled.div`
```
Comment (amina-deriv) [inline src/pages/unsubscribe/index.tsx:13]: Pls fix the typo` UnsubscrubeWrapper`

File: src/pages/unsubscribe/index.tsx [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                }
+            },
+        )
+    }
+    return (
+        <Layout>
+            {loading && (
+                <UnsubscrubeWrapper>
+                    <Spinner />
+                </UnsubscrubeWrapper>
```
Comment (amina-deriv) [inline src/pages/unsubscribe/index.tsx:157]: here too

File: src/pages/unsubscribe/index.tsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            )}
+            {!loading && (
+                <UnsubscrubeWrapper>
```
Comment (amina-deriv) [inline src/pages/unsubscribe/index.tsx:160]: here too

File: src/pages/unsubscribe/index.tsx [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                            </ConfirmWrapper>
+                        </UnsubscribeForm>
+                    )}
+                </UnsubscrubeWrapper>
+            )}
+        </Layout>
+    )
+}
+
+export default WithIntl()(UnsubscrubePage)
```
Comment (amina-deriv) [inline src/pages/unsubscribe/index.tsx:189]: here too

File: src/common/utility.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (prince-deriv) [reply src/common/utility.ts:358]: nope

File: src/pages/unsubscribe/index.tsx (hunk 2 above)
Comment (NikitK-deriv) [reply src/pages/unsubscribe/index.tsx:13]: Could you share what kind of errors you noticed @amina-deriv? Unfortunately locally I haven't any

Comment (amina-deriv) [reply src/pages/unsubscribe/index.tsx:13]: i mean the spelling `UnsubscrubeWrapper` . It should be `UnsubscribeWrapper`. right?

File: src/common/utility.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (NikitK-deriv) [reply src/common/utility.ts]: Resolved, thanks

File: src/pages/unsubscribe/index.tsx (hunk 2 above)
Comment (NikitK-deriv) [reply src/pages/unsubscribe/index.tsx:13]: Ohh my 😄 Should be, thanks

Comment (mohammad-rework) [reply src/pages/unsubscribe/index.tsx:13]: I think it's not resolved yet right?

File: src/pages/unsubscribe/index.tsx [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+                if (!response.error) {
+                    setLoading(false)
+                    setCompleteStatus(true)
+                }
+                if (response.error) {
+                    navigate('https://app.deriv.com/account/personal-details')
+                }
+            },
+        )
+    }
```
Comment (michio-rework) [inline src/pages/unsubscribe/index.tsx:152]: please wrap this up with `useCallback` if it's possible thanks :pray:

File: src/pages/unsubscribe/index.tsx [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                <UnsubscribeWrapper>
+                    <Spinner />
+                </UnsubscribeWrapper>
+            )}
+            {!loading && (
+                <UnsubscribeWrapper>
+                    {complete_status ? (
+                        <SuccessCard>
+                            <img src={CheckIcon} alt="sucess" width={48} height={48} />
+                            Unsubscribe successfully
```
Comment (michio-rework) [inline src/pages/unsubscribe/index.tsx:165]: don't we need translation for this string? 
I'm not sure, you can just ignore this if it's irrelevant.

File: src/common/url-base64-functions.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+const isValid = (src: any) => {
+    if (typeof src !== 'string') return false
+    const s = src?.replace(/\s+/g, '')?.replace(/={0,2}$/, '')
+    return !/[^\s0-9a-zA-Z\+/]/.test(s) || !/[^\s0-9a-zA-Z\-_]/.test(s)
+}
+export { decode as fromBase64 }
+export { encode as toBase64 }
+export { encode }
+export { decode }
+export { isValid }
```
Comment (michio-rework) [inline src/common/url-base64-functions.ts:168]: can you please provide short simple `jsDoc` for the functions as well?
it's gonna be hard for people to understand the whole module and what it does in the future. 
also the namings of the functions and variables can be much better, and please add mdn links for `atob` ,`btoa` functions. thanks thanks :pray:

**Review by michio-rework — CHANGES_REQUESTED** [review]
please check the comments :pray:

File: src/pages/unsubscribe/index.tsx (hunk 7 above)
Comment (prince-deriv) [reply src/pages/unsubscribe/index.tsx:165]: we need translation for this, This is overlooked thanks @michio-rework

File: src/pages/unsubscribe/index.tsx [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    const [complete_status, setCompleteStatus] = useState(false)
+    const [loading, setLoading] = useState(false)
+
+    const query = queryParams.get('hash') || ''
+    const unsubscribe_hash = decode(query).split('+')
+    const binary_user_id = unsubscribe_hash[0]
+    const checksum = unsubscribe_hash[1]
+
+    const deriv_api = useDerivApi()
+    const { send } = deriv_api
```
Comment (prince-deriv) [inline src/pages/unsubscribe/index.tsx:132]: use the one from the store so it will not open a new websocket connection.

Comment (NikitK-deriv) [reply src/pages/unsubscribe/index.tsx:132]: Done, thanks for note

File: src/pages/unsubscribe/index.tsx (hunk 7 above)
Comment (NikitK-deriv) [reply src/pages/unsubscribe/index.tsx:165]: Done 🙇

File: src/pages/unsubscribe/index.tsx (hunk 6 above)
Comment (NikitK-deriv) [reply src/pages/unsubscribe/index.tsx]: Possible, thanks for not. Thought it's not very necessarry in little page case

File: src/pages/unsubscribe/index.tsx [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                checksum: checksum,
+            },
+            (response) => {
+                if (!response.error) {
+                    setLoading(false)
+                    setCompleteStatus(true)
+                }
+                if (response.error) {
+                    navigate('https://app.deriv.com/account/personal-details')
+                }
```
Comment (prince-deriv) [inline src/pages/unsubscribe/index.tsx:148]: can we use `if` and `else` instead?

Comment (NikitK-deriv) [reply src/pages/unsubscribe/index.tsx:148]: Resolved

File: src/common/url-base64-functions.ts [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+// takes Base 64 string and returns encrypted string
+const encode = (src: string, urlsafe = false) =>
+    urlsafe ? defendURI(encodeString(src)) : encodeString(src)
+
+// takes encrypted string and returns decoded Base 64 string
+const decode = (src: string) => toUTF8(_unURI(src))
+
+// checks that string is valid for decode
+const isValid = (src: any) => {
```
Comment (michio-rework) [inline src/common/url-base64-functions.ts:185]:
```suggestion
/**
 * @description checks that string is valid for decode,if type of is not string, validation will not pass
 * @param src if type of is not string, validation will not pass, this function only validates string params
 * @returns {boolean} the validity of the input string based on the provided regex
 */
const isValid = (src: unknown) => {
    if (typeof src !== 'string') return false
    const s = src?.replace(/\s+/g, '')?.replace(/={0,2}$/, '')
    return !/[^\s0-9a-zA-Z\+/]/.test(s) || !/[^\s0-9a-zA-Z\-_]/.test(s)
}
```
by documentation and comments I meant `jsDoc`, if you wanna more about it please check [this cheatsheet](https://devhints.io/jsdoc).

the `unknown` suggestion description: 
if you don't know the type of the argument, please use `unknown` but I think we can use `string` here, please check if you can remove the `any` :pray: 
please update other functions as well if it's makes sense ( some of the are just easy to understand ) 
if you had any questions, please let me know :pray:

Comment (NikitK-deriv) [reply src/common/url-base64-functions.ts]: Done @michio-rework please check is this lind of docks enough for us? 🙏

**Review by michio-rework — APPROVED** [review]

**Comment by michio-rework** [comment]
please fix the sonarcloud bug before merging @NikitK-deriv , thanks thanks :pray:

**Review by prince-deriv — APPROVED** [review]

**Review by prince-deriv — APPROVED** [review]

**Pull request closed without merging**